File: 链表/2.两数相加.py

# 遍历

class Solution1:
    def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        # 注意进位是十位, 使用 // floor division 获得商
        dummy = ListNode()
        cur = dummy
        carry = 0
        while l1 or l2 or carry:
            sum_ = carry
            if l1:
                sum_ += l1.val
                l1 = l1.next
            if l2:
                sum_ += l2.val
                l2 = l2.next
            cur.next = ListNode(sum_ % 10)
            carry = sum_ // 10
            cur = cur.next
        return dummy.next

File: 链表/2.两数相加.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} l1
 * @param {ListNode} l2
 * @return {ListNode}
 */
var addTwoNumbers = function (l1, l2) {
  const dummy = new ListNode(0);
  let cur = dummy;
  let sum,
    carry = 0;
  // 从低位加到高位
  while (l1 || l2) {
    sum = carry;
    if (l1) {
      sum += l1.val;
      l1 = l1.next;
    }
    if (l2) {
      sum += l2.val;
      l2 = l2.next;
    }
    carry = Math.floor(sum / 10);
    cur.next = new ListNode(sum % 10);
    cur = cur.next;
  }
  cur.next = carry ? new ListNode(carry) : null;
  return dummy.next;
};
